Added possibility to get view types from the adapter to add support for GridLayoutManager.SpanSizeLookup

meruiden opened the pull request:
Example usage:

```kotlin
private val loadingIndicatorType = adapter.getItemViewTypeFromClass(LoadingIndicator::class.java)

override fun getSpanSize(position: Int): Int {
    val itemType = adapter.getItemViewType(position)

    return if(itemType == loadingIndicatorType) {
        spanCount
    } else {
        1
    }
}
```

See #15

meruiden committed:
Added possibility to get view types from the adapter to add support for GridLayoutManager.SpanSizeLookup

meruiden committed:
Moved InternalHolders to ExternalHolders to expose these classes to be used in GridLayoutManager.SpanSizeLookup

meruiden committed:
Changed getCreatorUniqueIndex type to any type to allow using this from classes rather than just instances

meruiden committed:
Removed iml file

idanatz commented on oneadapter/src/main/java/com/idanatz/oneadapter/internal/holders_creators/ViewHolderCreatorsStore.kt:
```diff
@@ -11,7 +11,7 @@ internal class ViewHolderCreatorsStore {
         holderCreators[clazz] = creator
     }
 
-    fun getCreatorUniqueIndex(clazz: Class<Diffable>): Int {
```
What is the reason for this change?
Each item type the adapter holds must implement Diffable, it doesn't matter if its an internal holder type or external class from the user.

meruiden replied:
The reason I changed this is to allow calling the function like this:

```kotlin
getCreatorUniqueIndex(LoadingIndicator::class.java)
```

When the type is < Diffable > rather then <*> the compiler complains that LoadingIndicator is not a Diffable even though it is. This does work with instances but not when using ``` LoadingIndicator::class.java ```. You could do a runtime check maybe and throw if it doesn't inherit from Diffable.

idanatz replied:
ViewHolderCreatorsStore is an internal class of the library, it should not be used with classes other than DIffable.

The way i see it, the function on OneAdapter.kt can be called with type < * >, but inside InternalAdapter, the function getItemViewTypeFromClass will validate the class if it is of type DIffable (using the Validator class) and throw a new exception (that will be written in the Exceptions.kt)

class UnsupportedClassException(msg: String = "Class must implement Diffable interface") : Throwable(msg)

meruiden replied:
I pushed the changes, let me know what u think :)

idanatz commented on oneadapter/src/main/java/com/idanatz/oneadapter/internal/InternalAdapter.kt:
```diff
@@ -352,6 +352,10 @@ internal class InternalAdapter(val recyclerView: RecyclerView) : RecyclerView.Ad
     }
     //endregion
 
+    fun getItemViewTypeFromClass(clazz: Class<*>): Int? {
```
change to Class < Diffable > for the reason written in 
ViewHolderCreatorsStore.kt

idanatz replied:
and place it below 
fun getItemViewType

idanatz commented on oneadapter/src/main/java/com/idanatz/oneadapter/external/holders/ExternalHolders.kt:
```diff
@@ -1,15 +1,15 @@
 import java.util.*
 
-internal abstract class InternalHolderModel : Diffable {
+abstract class ExternalHolderModel : Diffable {
```
External has no clear meaning... 
If these classes should be exposed to the user than change the base class name from InternalHolderModel to OneHolderModel as its a general public holder model of OneAdapter

meruiden replied:
Makes sense, il change it to OneHolderModel then. should I keep the new folder structure or put it back in internal/holders?

idanatz commented on oneadapter/src/main/java/com/idanatz/oneadapter/OneAdapter.kt:
```diff
@@ -79,6 +79,14 @@ class OneAdapter(recyclerView: RecyclerView) {
         }
     }
 
+    fun getItemViewType(position: Int): Int {
```
change to Class < Diffable > for the reason written in
ViewHolderCreatorsStore.kt

idanatz replied:
and place these functions at the end of the file

idanatz reviewed (changes requested):
Thanks for the PR!
Please see my comments

Commit:
fixed WhenReachingThresholdOnBindShouldBeCalledOnce.kt wrong logic
changed some leftover Class<*> to Class<Diffable>
added some extension functions to item list

meruiden committed:
Reformat code and follow naming convention

meruiden committed:
Change getCreatorUniqueIndex back to how it was and add type check to getItemViewTypeFromClass

idanatz commented:
Looks good.

a couple of final thins:
1. Before addressing the next bullets, pull from develop branch, I've made local changes before your PR and pushed them a few hours ago, after that address the issues below.

2. I know I wrote it but on seconds thought,  change 
class UnsupportedClassException(msg: String = "Class must implement Diffable interface") : Throwable(msg)
to 
class UnsupportedClassException(msg: String = "") : Throwable(msg)
and move the msg "Class must implement Diffable interface" to the throw site.
it's more understandable when reading the code to see the actual message.

3. Why is 
fun getItemViewTypeFromClass(clazz: Class<*>): Int?
returning Int? ? getCreatorUniqueIndex returns Int, it will never be nullable.

4. add this documentation below
     /**
     * Retrieves the view type of an item with a given class.
     * Note that this class must implement the Diffable interface and the adapter must contain items of that class.
     * @throws UnsupportedClassException if the class does not implement the Diffable interface.
     */
fun getItemViewTypeFromClass(clazz: Class<*>): Int?

Note that there are some validations that need to be implemented now that the getItemViewType and getCreatorUniqueIndex are publicly accessible, but I will add them after this PR will be merged.

meruiden committed:
Merge with master

meruiden commented:
>Why is
>fun getItemViewTypeFromClass(clazz: Class<*>): Int?
>returning Int? ? getCreatorUniqueIndex returns Int, it will never be nullable.

Whoops that shouldn't be nullable indeed. I'll remove this

meruiden committed:
After pulling the develop branch from the original source, I added the changes (some small refactoring) requested from the maintainer

meruiden commented:
> Looks good.
> 
> a couple of final thins:
> 
> 1. Before addressing the next bullets, pull from develop branch, I've made local changes before your PR and pushed them a few hours ago, after that address the issues below.
> 2. I know I wrote it but on seconds thought,  change
>    class UnsupportedClassException(msg: String = "Class must implement Diffable interface") : Throwable(msg)
>    to
>    class UnsupportedClassException(msg: String = "") : Throwable(msg)
>    and move the msg "Class must implement Diffable interface" to the throw site.
>    it's more understandable when reading the code to see the actual message.
> 3. Why is
>    fun getItemViewTypeFromClass(clazz: Class<*>): Int?
>    returning Int? ? getCreatorUniqueIndex returns Int, it will never be nullable.
> 4. add this documentation below
>    /**
>    
>    * Retrieves the view type of an item with a given class.
>    * Note that this class must implement the Diffable interface and the adapter must contain items of that class.
>    * @throws UnsupportedClassException if the class does not implement the Diffable interface.
>      _/
>      fun getItemViewTypeFromClass(clazz: Class<_>): Int?
> 
> Note that there are some validations that need to be implemented now that the getItemViewType and getCreatorUniqueIndex are publicly accessible, but I will add them after this PR will be merged.

I pulled the latest changes from develop and changed the stuff you requested, merged all in to master and pushed.

idanatz commented:
merged your PR, thanks!
will be released as a minor version in a few days after il check everything is good.
